src/fb.c

```diff
@@ -1,63 +1,275 @@
+static void fb_console_blit_all(void)
 {
-    uint32_t *p = data;
+    for (u32 row = 0; row < console.n_rows; ++row)
+        fb_console_blit_row(row);
+}
 
-    for (int i = 0; i < h; i++)
-        for (int j = 0; j < w; j++)
-            fb[x + j + (y + i) * fb_s] = rgbx_to_rgb30(p[j + i * stride]);
+static void fb_console_putbyte(u8 c)
+{
+    if (console.cursor.col >= console.n_cols)
+        return;
+
+    console.text[fb_console_get_row(console.cursor.row)][console.cursor.col] = c;
+    fb_console_blit_char(console.cursor);
+
+    console.cursor.col++;
+}
+
+static void fb_check_scroll(void)
+{
+    if (console.cursor.row < console.n_rows)
+        return;
+
+    memset(console.text[fb_console_get_row(0)], 0, CONSOLE_MAX_COLS);
+    console.row_offset = (console.row_offset + 1) % console.n_rows;
+    console.cursor.row--;
+
+    assert(console.cursor.row < console.n_rows);
+
+    fb_console_blit_all();
+}
+
+void fb_console_putc(u8 c)
+{
+    if (!smp_is_primary())
+        return;
+
+    if (c == '\n') {
+        console.cursor.row++;
+        console.cursor.col = 0;
+    } else if (c == '\r') {
+        console.cursor.col = 0;
+    } else if (c >= 0x20 && c <= 0x7e) {
+        fb_console_putbyte(c);
+    } else {
+        fb_console_putbyte('?');
+    }
+
+    fb_check_scroll();
+}
+
+void fb_console_write(const void *buf, size_t count)
+{
+    const u8 *p = buf;
+
+    while (count--)
+        fb_console_putc(*p++);
+}
+
+void fb_console_init(void)
+{
+    if (cur_boot_args.video.depth & FB_DEPTH_FLAG_RETINA) {
+        console.font.ptr = _binary_build_font_retina_bin_start;
+        console.font.width = 16;
+        console.font.height = 32;
+    } else {
+        console.font.ptr = _binary_build_font_bin_start;
+        console.font.width = 8;
+        console.font.height = 16;
+    }
+
+    console.n_rows =
+        min(CONSOLE_MAX_ROWS,
+            (fb.height - CONSOLE_MARGIN_TOP - CONSOLE_MARGIN_BOTTOM) / console.font.height);
+    console.n_cols =
+        min(CONSOLE_MAX_COLS, (fb.width - CONSOLE_MARGIN_LEFT - CONSOLE_MARGIN_RIGHT - logo.width) /
+                                  console.font.width);
+
+    if (console.row_offset > 0) {
+        /*
+         * TODO: handle this case better if we ever print more than CONSOLE_MAX_ROWS before calling
+         * fb_console_init
+         */
+        printf("fb console: printed more than %d rows before init, discarding early console.\n",
+               CONSOLE_MAX_ROWS);
+        memset(console.text, 0, sizeof(console.text));
+        console.cursor.col = 0;
+        console.cursor.row = 0;
+    }
+
+    /*
+     * this size of our console may have shrunk and we'll need to throw away some early rows
+     * and copy later ones to the start.
+     * we can't use fb_check_scroll() here because we will need to copy rows that are already
+     * larger than console.n_rows.
+     */
+    if (console.cursor.row > console.n_rows) {
```

Comment:
This works fine the way it is. `fb_check_scroll` is meant to restore the invariant `row < n_rows` when being called with `row <= n_rows`.
It will do `n_row--` before hitting that assert. 

I could make this `>=` and add a `+1/-1` in some of the lines below but I think it's easier to read this way. No strong opinion though.

Reply:
I was encountering an issue where that assert was getting triggered. You're right that the case I mention shouldn't actually happen and I think maybe it was just a coincidence or something else was off. So this is fine.

Reply:
Did you maybe by chance run an outdated version? I accidentally removed the `fb_check_scroll` call while rebasing and then later ran into that assert because `row` was `n_rows + 1`.

Reply:
I bet that's what happened. I was using a very weird merge briefly and I thought I had reset everything clean but I think I ended up with thing in a weird state.